{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "45967421-0ab1-4269-b6f2-0dccc2d9a73f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# retrieve plasmid sequence from GenBank\n",
    "from pydna.genbank import Genbank\n",
    "from Bio import SeqIO\n",
    "from Bio.Restriction import AllEnzymes\n",
    "from pydna.parsers import parse_snapgene\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import itertools\n",
    "\n",
    "from snapgene_reader import snapgene_file_to_dict, snapgene_file_to_seqrecord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f960e634-8ae1-433b-a443-2636a5ecebf4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "EcoRI"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "AllEnzymes.get(\"EcoRI\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5cf87e57-70a1-4f04-8a3d-8802d9f4e444",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pydna.dseqrecord import Dseqrecord\n",
    "from pydna.design import primer_design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "14956dc0-e335-4341-80cc-8103659129f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_from_genebank(email, sequence):\n",
    "    gb = Genbank(users_email=email)  \n",
    "    seq = gb.nucleotide(sequence)   \n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "df90d57a-9ebf-4c1e-8a19-692332fa83a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "email = \"mikelgarciadata@gmail.com\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cd239255-8e66-4bfb-a34f-f37982229b20",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Some gene sequence\n",
    "seqs_id = [\"M61151.1\", \"X51424.1\", \"PV006683.1\", \"PV006682.1\", \"OR455449.1\"]\n",
    "# Create a Dseq object\n",
    "seqs = []\n",
    "for seq in seqs_id:\n",
    "    seqs.append(load_from_genebank(email, seq))\n",
    "\n",
    "genes = list(map(lambda seq: Dseqrecord(seq), seqs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6a469c22-ca79-4752-977f-d8855e5a3e43",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define primer length (typical range 18-22 nt)\n",
    "primer_length = 20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "738e40ab-2010-46ca-afb3-1bfbef66a66f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Forward primer: first 20 nt\n",
    "fwd_primers = list(map(lambda gene: gene[:primer_length], genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d4050b7d-9fd0-4f99-b598-10df9ee3c86e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reverse primer: last 20 nt, reverse complement\n",
    "rev_primers = list(map(lambda gene: gene[-primer_length:].rc(), genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1acccff9-792a-4aa2-9859-981658181038",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward primer: Dseqrecord\n",
      "circular: False\n",
      "size: 20\n",
      "ID: part_M611511\n",
      "Name: part_ATUAUX\n",
      "Description: Agrobacterium rhizogenes trytophan oxygenase (aux1) and hydrolase (aux2) genes, complete cds\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=23-AUG-1993\n",
      "/accessions=['M61151', 'J03688']\n",
      "/sequence_version=1\n",
      "/keywords=['aux1 gene', 'aux2 gene', 'hydrolase', 'tryptophan oxygenase']\n",
      "/source=Rhizobium rhizogenes (Agrobacterium rhizogenes)\n",
      "/organism=Rhizobium rhizogenes\n",
      "/taxonomy=['Bacteria', 'Pseudomonadati', 'Pseudomonadota', 'Alphaproteobacteria', 'Hyphomicrobiales', 'Rhizobiaceae', 'Rhizobium/Agrobacterium group', 'Rhizobium']\n",
      "/references=[Reference(title='The TR.DNA region carrying the auxin synthesis genes of the Agrobacterium rhizogenes agropine-type pRiA4: Nucleotide sequence analysis and introduction into tobacco plants', ...)]\n",
      "Dseq(-20)\n",
      "GTCGACAGTCGCAACAGCAA\n",
      "CAGCTGTCAGCGTTGTCGTT\n",
      "Reverse primer: Dseqrecord\n",
      "circular: False\n",
      "size: 20\n",
      "ID: part_M611511_rc\n",
      "Name: part_ATUAUX_rc\n",
      "Description: Agrobacterium rhizogenes trytophan oxygenase (aux1) and hydrolase (aux2) genes, complete cds_rc\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "Dseq(-20)\n",
      "GTCGACGTTTGCCGGCGAGG\n",
      "CAGCTGCAAACGGCCGCTCC\n",
      "Forward primer: Dseqrecord\n",
      "circular: False\n",
      "size: 20\n",
      "ID: part_X514241\n",
      "Name: part_X51424\n",
      "Description: Bacillus subtilis cwlA gene for cell wall hydrolase\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=23-OCT-2008\n",
      "/accessions=['X51424', 'M60054']\n",
      "/sequence_version=1\n",
      "/keywords=['cell wall hydrolase', 'cwlA gene', 'hydrolase']\n",
      "/source=Bacillus subtilis\n",
      "/organism=Bacillus subtilis\n",
      "/taxonomy=['Bacteria', 'Bacillati', 'Bacillota', 'Bacilli', 'Bacillales', 'Bacillaceae', 'Bacillus']\n",
      "/references=[Reference(title='Cloning, sequencing and genetic mapping of a Bacillus subtilis cell wall hydrolase gene', ...), Reference(title='Direct Submission', ...)]\n",
      "Dseq(-20)\n",
      "GATCGATTACAAACAATTGA\n",
      "CTAGCTAATGTTTGTTAACT\n",
      "Reverse primer: Dseqrecord\n",
      "circular: False\n",
      "size: 20\n",
      "ID: part_X514241_rc\n",
      "Name: part_X51424_rc\n",
      "Description: Bacillus subtilis cwlA gene for cell wall hydrolase_rc\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "Dseq(-20)\n",
      "GCAGATGTTGCTAAAAGTAA\n",
      "CGTCTACAACGATTTTCATT\n",
      "Forward primer: Dseqrecord\n",
      "circular: False\n",
      "size: 20\n",
      "ID: part_PV0066831\n",
      "Name: part_PV006683\n",
      "Description: Burkholderia cepacia strain ZY1 hydrolase (hyd4173) gene, complete cds\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=18-FEB-2025\n",
      "/accessions=['PV006683']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Burkholderia cepacia\n",
      "/organism=Burkholderia cepacia\n",
      "/taxonomy=['Bacteria', 'Pseudomonadati', 'Pseudomonadota', 'Betaproteobacteria', 'Burkholderiales', 'Burkholderiaceae', 'Burkholderia', 'Burkholderia cepacia complex']\n",
      "/references=[Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=defaultdict(<class 'dict'>, {'Assembly-Data': {'Sequencing Technology': 'Sanger dideoxy sequencing'}})\n",
      "Dseq(-20)\n",
      "GTGTCGGTCATCGAATCCAC\n",
      "CACAGCCAGTAGCTTAGGTG\n",
      "Reverse primer: Dseqrecord\n",
      "circular: False\n",
      "size: 20\n",
      "ID: part_PV0066831_rc\n",
      "Name: part_PV006683_rc\n",
      "Description: Burkholderia cepacia strain ZY1 hydrolase (hyd4173) gene, complete cds_rc\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "Dseq(-20)\n",
      "TCAGCGCGCGCCGGACAGAT\n",
      "AGTCGCGCGCGGCCTGTCTA\n",
      "Forward primer: Dseqrecord\n",
      "circular: False\n",
      "size: 20\n",
      "ID: part_PV0066821\n",
      "Name: part_PV006682\n",
      "Description: Burkholderia cepacia strain ZY1 hydrolase (hyd2191) gene, complete cds\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=18-FEB-2025\n",
      "/accessions=['PV006682']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Burkholderia cepacia\n",
      "/organism=Burkholderia cepacia\n",
      "/taxonomy=['Bacteria', 'Pseudomonadati', 'Pseudomonadota', 'Betaproteobacteria', 'Burkholderiales', 'Burkholderiaceae', 'Burkholderia', 'Burkholderia cepacia complex']\n",
      "/references=[Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=defaultdict(<class 'dict'>, {'Assembly-Data': {'Sequencing Technology': 'Sanger dideoxy sequencing'}})\n",
      "Dseq(-20)\n",
      "ATGGCGCACGGGCTGGGTGG\n",
      "TACCGCGTGCCCGACCCACC\n",
      "Reverse primer: Dseqrecord\n",
      "circular: False\n",
      "size: 20\n",
      "ID: part_PV0066821_rc\n",
      "Name: part_PV006682_rc\n",
      "Description: Burkholderia cepacia strain ZY1 hydrolase (hyd2191) gene, complete cds_rc\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "Dseq(-20)\n",
      "TTATAAGTCGACATGCGCGT\n",
      "AATATTCAGCTGTACGCGCA\n",
      "Forward primer: Dseqrecord\n",
      "circular: False\n",
      "size: 20\n",
      "ID: part_OR4554491\n",
      "Name: part_OR455449\n",
      "Description: Rhodococcus sp. D-6 hydrolase (ipcH) gene, partial cds\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "/topology=linear\n",
      "/data_file_division=BCT\n",
      "/date=26-JUN-2024\n",
      "/accessions=['OR455449']\n",
      "/sequence_version=1\n",
      "/keywords=['']\n",
      "/source=Rhodococcus sp. D-6\n",
      "/organism=Rhodococcus sp. D-6\n",
      "/taxonomy=['Bacteria', 'Bacillati', 'Actinomycetota', 'Actinomycetes', 'Mycobacteriales', 'Nocardiaceae', 'Rhodococcus']\n",
      "/references=[Reference(title='The novel hydrolase IpcH responsible for the initial isoprocarb degradation step in Rhodococcus sp. D-6', ...), Reference(title='Direct Submission', ...)]\n",
      "/structured_comment=defaultdict(<class 'dict'>, {'Assembly-Data': {'Sequencing Technology': 'Sanger dideoxy sequencing'}})\n",
      "Dseq(-20)\n",
      "ATGTGTTGCGAAGATCACCT\n",
      "TACACAACGCTTCTAGTGGA\n",
      "Reverse primer: Dseqrecord\n",
      "circular: False\n",
      "size: 20\n",
      "ID: part_OR4554491_rc\n",
      "Name: part_OR455449_rc\n",
      "Description: Rhodococcus sp. D-6 hydrolase (ipcH) gene, partial cds_rc\n",
      "Number of features: 0\n",
      "/molecule_type=DNA\n",
      "Dseq(-20)\n",
      "TCATCCTGTCATGTCCCTGC\n",
      "AGTAGGACAGTACAGGGACG\n"
     ]
    }
   ],
   "source": [
    "for fwd_primer, rev_primer in zip(fwd_primers, rev_primers):\n",
    "    print(\"Forward primer:\", fwd_primer)\n",
    "    print(\"Reverse primer:\", rev_primer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b6ad722d-05ee-4c0a-8cc0-5f4083425af6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\u001b[31mSignature:\u001b[39m\n",
       "primer_design(\n",
       "    template,\n",
       "    fp=\u001b[38;5;28;01mNone\u001b[39;00m,\n",
       "    rp=\u001b[38;5;28;01mNone\u001b[39;00m,\n",
       "    limit=\u001b[32m13\u001b[39m,\n",
       "    target_tm=\u001b[32m55.0\u001b[39m,\n",
       "    tm_func=<function tm_default at \u001b[32m0x7fee166e1b20\u001b[39m>,\n",
       "    estimate_function=\u001b[38;5;28;01mNone\u001b[39;00m,\n",
       "    **kwargs,\n",
       ")\n",
       "\u001b[31mSource:\u001b[39m   \n",
       "\u001b[38;5;28;01mdef\u001b[39;00m primer_design(\n",
       "    template,\n",
       "    fp=\u001b[38;5;28;01mNone\u001b[39;00m,\n",
       "    rp=\u001b[38;5;28;01mNone\u001b[39;00m,\n",
       "    limit=\u001b[32m13\u001b[39m,\n",
       "    target_tm=\u001b[32m55.0\u001b[39m,\n",
       "    tm_func=_tm_default,\n",
       "    estimate_function=\u001b[38;5;28;01mNone\u001b[39;00m,\n",
       "    **kwargs,\n",
       "):\n",
       "    \u001b[33m\"\"\"This function designs a forward primer and a reverse primer for PCR amplification\u001b[39m\n",
       "\u001b[33m    of a given template sequence.\u001b[39m\n",
       "\n",
       "\u001b[33m    The template argument is a Dseqrecord object or equivalent containing the template sequence.\u001b[39m\n",
       "\n",
       "\u001b[33m    The optional fp and rp arguments can contain an existing primer for the sequence (either the forward or reverse primer).\u001b[39m\n",
       "\u001b[33m    One or the other primers can be specified, not both (since then there is nothing to design!, use the pydna.amplify.pcr function instead).\u001b[39m\n",
       "\n",
       "\u001b[33m    The limit argument is the minimum length of the primer. The default value is 13.\u001b[39m\n",
       "\n",
       "\u001b[33m    If one of the primers is given, the other primer is designed to match in terms of Tm.\u001b[39m\n",
       "\u001b[33m    If both primers are designed, they will be designed to target_tm\u001b[39m\n",
       "\n",
       "\u001b[33m    tm_func is a function that takes an ascii string representing an oligonuceotide as argument and returns a float.\u001b[39m\n",
       "\u001b[33m    Some useful functions can be found in the :mod:`pydna.tm` module, but can be substituted for a custom made function.\u001b[39m\n",
       "\n",
       "\u001b[33m    estimate_function is a tm_func-like function that is used to get a first guess for the primer design, that is then used as starting\u001b[39m\n",
       "\u001b[33m    point for the final result. This is useful when the tm_func function is slow to calculate (e.g. it relies on an\u001b[39m\n",
       "\u001b[33m    external API, such as the NEB primer design API). The estimate_function should be faster than the tm_func function.\u001b[39m\n",
       "\u001b[33m    The default value is `None`.\u001b[39m\n",
       "\u001b[33m    To use the default `tm_func` as estimate function to get the NEB Tm faster, you can do:\u001b[39m\n",
       "\u001b[33m    `primer_design(dseqr, target_tm=55, tm_func=tm_neb, estimate_function=tm_default)`.\u001b[39m\n",
       "\n",
       "\u001b[33m    The function returns a pydna.amplicon.Amplicon class instance. This object has\u001b[39m\n",
       "\u001b[33m    the object.forward_primer and object.reverse_primer properties which contain the designed primers.\u001b[39m\n",
       "\n",
       "\n",
       "\u001b[33m    Parameters\u001b[39m\n",
       "\u001b[33m    ----------\u001b[39m\n",
       "\n",
       "\u001b[33m    template : pydna.dseqrecord.Dseqrecord\u001b[39m\n",
       "\u001b[33m        a Dseqrecord object. The only required argument.\u001b[39m\n",
       "\n",
       "\u001b[33m    fp, rp : pydna.primer.Primer, optional\u001b[39m\n",
       "\u001b[33m        optional pydna.primer.Primer objects containing one primer each.\u001b[39m\n",
       "\n",
       "\u001b[33m    target_tm : float, optional\u001b[39m\n",
       "\u001b[33m        target tm for the primers, set to 55Â°C by default.\u001b[39m\n",
       "\n",
       "\u001b[33m    tm_func : function\u001b[39m\n",
       "\u001b[33m        Function used for tm calculation. This function takes an ascii string\u001b[39m\n",
       "\u001b[33m        representing an oligonuceotide as argument and returns a float.\u001b[39m\n",
       "\u001b[33m        Some useful functions can be found in the :mod:`pydna.tm` module, but can be\u001b[39m\n",
       "\u001b[33m        substituted for a custom made function.\u001b[39m\n",
       "\n",
       "\u001b[33m    Returns\u001b[39m\n",
       "\u001b[33m    -------\u001b[39m\n",
       "\u001b[33m    result : Amplicon\u001b[39m\n",
       "\n",
       "\u001b[33m    Examples\u001b[39m\n",
       "\u001b[33m    --------\u001b[39m\n",
       "\n",
       "\u001b[33m    >>> from pydna.dseqrecord import Dseqrecord\u001b[39m\n",
       "\u001b[33m    >>> t=Dseqrecord(\"atgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatg\")\u001b[39m\n",
       "\u001b[33m    >>> t\u001b[39m\n",
       "\u001b[33m    Dseqrecord(-64)\u001b[39m\n",
       "\u001b[33m    >>> from pydna.design import primer_design\u001b[39m\n",
       "\u001b[33m    >>> ampl = primer_design(t)\u001b[39m\n",
       "\u001b[33m    >>> ampl\u001b[39m\n",
       "\u001b[33m    Amplicon(64)\u001b[39m\n",
       "\u001b[33m    >>> ampl.forward_primer\u001b[39m\n",
       "\u001b[33m    f64 17-mer:5'-atgactgctaacccttc-3'\u001b[39m\n",
       "\u001b[33m    >>> ampl.reverse_primer\u001b[39m\n",
       "\u001b[33m    r64 18-mer:5'-catcgtaagtttcgaacg-3'\u001b[39m\n",
       "\u001b[33m    >>> print(ampl.figure())\u001b[39m\n",
       "\u001b[33m    5atgactgctaacccttc...cgttcgaaacttacgatg3\u001b[39m\n",
       "\u001b[33m                         ||||||||||||||||||\u001b[39m\n",
       "\u001b[33m                        3gcaagctttgaatgctac5\u001b[39m\n",
       "\u001b[33m    5atgactgctaacccttc3\u001b[39m\n",
       "\u001b[33m     |||||||||||||||||\u001b[39m\n",
       "\u001b[33m    3tactgacgattgggaag...gcaagctttgaatgctac5\u001b[39m\n",
       "\u001b[33m    >>> pf = \"GGATCC\" + ampl.forward_primer\u001b[39m\n",
       "\u001b[33m    >>> pr = \"GGATCC\" + ampl.reverse_primer\u001b[39m\n",
       "\u001b[33m    >>> pf\u001b[39m\n",
       "\u001b[33m    f64 23-mer:5'-GGATCCatgactgct..ttc-3'\u001b[39m\n",
       "\u001b[33m    >>> pr\u001b[39m\n",
       "\u001b[33m    r64 24-mer:5'-GGATCCcatcgtaag..acg-3'\u001b[39m\n",
       "\u001b[33m    >>> from pydna.amplify import pcr\u001b[39m\n",
       "\u001b[33m    >>> pcr_prod = pcr(pf, pr, t)\u001b[39m\n",
       "\u001b[33m    >>> print(pcr_prod.figure())\u001b[39m\n",
       "\u001b[33m          5atgactgctaacccttc...cgttcgaaacttacgatg3\u001b[39m\n",
       "\u001b[33m                               ||||||||||||||||||\u001b[39m\n",
       "\u001b[33m                              3gcaagctttgaatgctacCCTAGG5\u001b[39m\n",
       "\u001b[33m    5GGATCCatgactgctaacccttc3\u001b[39m\n",
       "\u001b[33m           |||||||||||||||||\u001b[39m\n",
       "\u001b[33m          3tactgacgattgggaag...gcaagctttgaatgctac5\u001b[39m\n",
       "\u001b[33m    >>> print(pcr_prod.seq)\u001b[39m\n",
       "\u001b[33m    GGATCCatgactgctaacccttccttggtgttgaacaagatcgacgacatttcgttcgaaacttacgatgGGATCC\u001b[39m\n",
       "\u001b[33m    >>> from pydna.primer import Primer\u001b[39m\n",
       "\u001b[33m    >>> pf = Primer(\"atgactgctaacccttccttggtgttg\", id=\"myprimer\")\u001b[39m\n",
       "\u001b[33m    >>> ampl = primer_design(t, fp = pf)\u001b[39m\n",
       "\u001b[33m    >>> ampl.forward_primer\u001b[39m\n",
       "\u001b[33m    myprimer 27-mer:5'-atgactgctaaccct..ttg-3'\u001b[39m\n",
       "\u001b[33m    >>> ampl.reverse_primer\u001b[39m\n",
       "\u001b[33m    r64 32-mer:5'-catcgtaagtttcga..atc-3'\u001b[39m\n",
       "\u001b[33m    \"\"\"\u001b[39m\n",
       "\n",
       "    \u001b[38;5;28;01mdef\u001b[39;00m design(target_tm, template):\n",
       "        \u001b[38;5;28;01mif\u001b[39;00m estimate_function:\n",
       "            first_guess = _design_primer(target_tm, template, limit, estimate_function)\n",
       "            \u001b[38;5;28;01mreturn\u001b[39;00m _design_primer(target_tm, template, limit, tm_func, len(first_guess))\n",
       "        \u001b[38;5;28;01melse\u001b[39;00m:\n",
       "            \u001b[38;5;28;01mreturn\u001b[39;00m _design_primer(target_tm, template, limit, tm_func)\n",
       "\n",
       "    \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28;01mnot\u001b[39;00m fp \u001b[38;5;28;01mand\u001b[39;00m \u001b[38;5;28;01mnot\u001b[39;00m rp:\n",
       "        \u001b[38;5;66;03m# _module_logger.debug(\"no primer given, design forward primer:\")\u001b[39;00m\n",
       "        fp = _Primer((design(target_tm, template)))\n",
       "        target_tm = tm_func(str(fp.seq))\n",
       "        \u001b[38;5;66;03m# _module_logger.debug(\"no primer given, design reverse primer:\")\u001b[39;00m\n",
       "        rp = _Primer(design(target_tm, template.reverse_complement()))\n",
       "    \u001b[38;5;28;01melif\u001b[39;00m fp \u001b[38;5;28;01mand\u001b[39;00m \u001b[38;5;28;01mnot\u001b[39;00m rp:\n",
       "        \u001b[38;5;28;01mtry\u001b[39;00m:\n",
       "            fp = _Anneal((fp,), template).forward_primers.pop()\n",
       "        \u001b[38;5;28;01mexcept\u001b[39;00m IndexError:\n",
       "            \u001b[38;5;28;01mraise\u001b[39;00m ValueError(\u001b[33m\"Forward primer does not anneal\"\u001b[39m)\n",
       "        \u001b[38;5;28;01mexcept\u001b[39;00m Exception:  \u001b[38;5;66;03m# pragma: no cover\u001b[39;00m\n",
       "            print(\u001b[33m\"Unexpected error\"\u001b[39m)\n",
       "            \u001b[38;5;28;01mraise\u001b[39;00m\n",
       "        target_tm = tm_func(fp.footprint)\n",
       "        \u001b[38;5;66;03m# _module_logger.debug(\"forward primer given, design reverse primer:\")\u001b[39;00m\n",
       "        rp = _Primer(design(target_tm, template.reverse_complement()))\n",
       "    \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28;01mnot\u001b[39;00m fp \u001b[38;5;28;01mand\u001b[39;00m rp:\n",
       "        \u001b[38;5;28;01mtry\u001b[39;00m:\n",
       "            rp = _Anneal((rp,), template).reverse_primers.pop()\n",
       "        \u001b[38;5;28;01mexcept\u001b[39;00m IndexError:\n",
       "            \u001b[38;5;28;01mraise\u001b[39;00m ValueError(\u001b[33m\"Reverse primer does not anneal\"\u001b[39m)\n",
       "        \u001b[38;5;28;01mexcept\u001b[39;00m Exception:  \u001b[38;5;66;03m# pragma: no cover\u001b[39;00m\n",
       "            print(\u001b[33m\"Unexpected error\"\u001b[39m)\n",
       "            \u001b[38;5;28;01mraise\u001b[39;00m\n",
       "        target_tm = tm_func(rp.footprint)\n",
       "        \u001b[38;5;66;03m# _module_logger.debug(\"reverse primer given, design forward primer:\")\u001b[39;00m\n",
       "        fp = _Primer(design(target_tm, template))\n",
       "    \u001b[38;5;28;01melse\u001b[39;00m:\n",
       "        \u001b[38;5;28;01mraise\u001b[39;00m ValueError(\u001b[33m\"Specify maximum one of the two primers.\"\u001b[39m)\n",
       "\n",
       "    \u001b[38;5;28;01mif\u001b[39;00m fp.id == \u001b[33m\"id\"\u001b[39m:  \u001b[38;5;66;03m# <unknown id>\u001b[39;00m\n",
       "        fp.id = \u001b[33m\"f{}\"\u001b[39m.format(len(template))\n",
       "\n",
       "    \u001b[38;5;28;01mif\u001b[39;00m rp.id == \u001b[33m\"id\"\u001b[39m:\n",
       "        rp.id = \u001b[33m\"r{}\"\u001b[39m.format(len(template))\n",
       "\n",
       "    \u001b[38;5;28;01mif\u001b[39;00m fp.name == \u001b[33m\"name\"\u001b[39m:\n",
       "        fp.name = \u001b[33m\"f{}\"\u001b[39m.format(len(template))\n",
       "\n",
       "    \u001b[38;5;28;01mif\u001b[39;00m rp.name == \u001b[33m\"name\"\u001b[39m:\n",
       "        rp.name = \u001b[33m\"r{}\"\u001b[39m.format(len(template))\n",
       "\n",
       "    fp.description = fp.id + \u001b[33m\" \"\u001b[39m + template.accession\n",
       "    rp.description = rp.id + \u001b[33m\" \"\u001b[39m + template.accession\n",
       "\n",
       "    ampl = _Anneal((fp, rp), template, limit=limit)\n",
       "\n",
       "    prod = ampl.products[\u001b[32m0\u001b[39m] \u001b[38;5;28;01mif\u001b[39;00m ampl.products \u001b[38;5;28;01melse\u001b[39;00m _Amplicon(\u001b[33m\"\"\u001b[39m)\n",
       "\n",
       "    \u001b[38;5;28;01mif\u001b[39;00m len(ampl.products) > \u001b[32m1\u001b[39m:\n",
       "        \u001b[38;5;28;01mimport\u001b[39;00m warnings \u001b[38;5;28;01mas\u001b[39;00m _warnings\n",
       "        \u001b[38;5;28;01mfrom\u001b[39;00m pydna \u001b[38;5;28;01mimport\u001b[39;00m _PydnaWarning\n",
       "\n",
       "        _warnings.warn(\n",
       "            \u001b[33m\"designed primers do not yield a unique PCR product\"\u001b[39m, _PydnaWarning\n",
       "        )\n",
       "\n",
       "    \u001b[38;5;28;01mreturn\u001b[39;00m prod\n",
       "\u001b[31mFile:\u001b[39m      ~/miniconda3/envs/pydnahk/lib/python3.13/site-packages/pydna/design.py\n",
       "\u001b[31mType:\u001b[39m      function"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "??primer_design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "93e51f71-9449-4201-972e-dbae536efdeb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Forward primer: GTCGACAGTCGCAACAGCAA\n",
      "Reverse primer: GTCGACGTTTGCCGGCGAGG\n",
      "Forward primer: GATCGATTACAAACAATTGAGAA\n",
      "Reverse primer: GCAGATGTTGCTAAAAGTAATT\n",
      "Forward primer: GTGTCGGTCATCGAATCCAC\n",
      "Reverse primer: TCAGCGCGCGCCGGACAGAT\n",
      "Forward primer: ATGGCGCACGGGCTGGGTGG\n",
      "Reverse primer: TTATAAGTCGACATGCGCGTCTCCTTGCTTCGCTGG\n",
      "Forward primer: ATGTGTTGCGAAGATCACCT\n",
      "Reverse primer: TCATCCTGTCATGTCCCTGC\n"
     ]
    }
   ],
   "source": [
    "# Or use Primer3 for more sophisticated design (optional):\n",
    "\n",
    "primers = list(map(lambda gene: primer_design(gene, target_tm=55, limit=20), genes))\n",
    "\n",
    "for primer in primers:\n",
    "    print(\"Forward primer:\", primer.forward_primer.seq)\n",
    "    print(\"Reverse primer:\", primer.reverse_primer.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f6e960a-1791-4fdf-97fe-9bf6d5aa4097",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
